Scripts/Runtime/Jobs/Collections/DynamicBufferSharedWriteController.cs

```diff
@@ -0,0 +1,335 @@
+        JobHandle GetSharedWriteDependency(SystemBase callingSystem, JobHandle callingSystemDependency);
+
+        /// <summary>
+        /// Sets the shared write dependency to the passed in <see cref="JobHandle"/>
+        /// </summary>
+        /// <param name="dependsOn">The <see cref="JobHandle"/> to set to</param>
+        void ForceSetSharedWriteDependency(JobHandle dependsOn);
+    }
+
+    /// <summary>
+    /// A utility class that handles managing access for shared writing (multiple job types writing at the same time)
+    /// so that jobs can be scheduled easily.
+    ///
+    /// ****IMPORTANT NOTE****
+    /// Using this class is very helpful for the specific task of writing to a <see cref="DynamicBuffer{T}"/> in
+    /// parallel from multiple job types. This class DOES NOT prevent multiple threads from writing to the same element.
+    /// Typically, this means that the use case should guarantee that there is no overlap in buffer element access.
+    /// For example, element ranges (1-n) are dedicated to specific worker indices.
+    /// (Worker 0 writes to element 0, Worker 3 writes to element 3, etc...)
+    ///
+    /// In order to achieve this functionality the class tracks the World's system order to the best of its ability.
+    /// There are cases where this isn't perfect and your won't get the results you expect.
+    /// These cases are outlined below to aid in debugging if needed.
+    ///
+    /// ALL OF THESE SHOULD BE VERY RARE
+    ///
+    /// - The <see cref="WorldCache"/> checks to see if it should be rebuilt by the number of systems in each
+    /// <see cref="ComponentSystemGroup"/>. It could be possible to reorder systems that touch your
+    /// <see cref="IBufferElementData"/> or have a combination of add/removes that keeps the count the same and thus
+    /// the <see cref="WorldCache"/> may not update properly and you will have to update manually
+    /// via <see cref="WorldCache.ForceRebuild"/>.
+    ///
+    /// - If you are using <see cref="EntityQuery"/>s that your Systems aren't aware of.
+    /// Ex. Queries created with <see cref="EntityManager.CreateEntityQuery"/> will not detect jobs that touch your
+    /// <see cref="IBufferElementData"/>. Always use the <see cref="SystemBase.GetEntityQuery"/> function.
+    ///
+    /// - If you are scheduling jobs outside of a System that operate on <see cref="IBufferElementData"/> this class
+    /// will not be aware of it and may lead to a dependency conflict. Consider using a
+    /// <see cref="CollectionAccessController{TContext}"/> to handle this instead.
+    ///
+    /// - If you are scheduling multiple jobs in a System that operates on the <see cref="IBufferElementData"/> where
+    /// one uses the Shared Write handle and the other(s) try and do an exclusive write or read, we treat that
+    /// System as only using the Shared Write handle.
+    ///
+    /// - If you are calling <see cref="SystemBase.GetEntityQuery"/> that operates on the
```

Comment:
If I understand correctly this is just sub-optimal scheduling for 1 frame. Right?